### PR TITLE
Make grpc reporter default and add retry

### LINKED ISSUE #1218: Use gRPC between collector and agent as default
We should decide when to use gRPC as default between agent and collector.

The first release with gRPC will be 1.9.0. If all goes well could we switch to the default in 1.10?

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #1218 #1229 

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

File: cmd/agent/app/reporter/grpc/collector_proxy.go [hunk 1]
```diff
@@ -49,10 +50,12 @@ func NewCollectorProxy(o *Options, mFactory metrics.Factory, logger *zap.Logger)
...
 		}
 		r.InitialAddrs(resolvedAddrs)
-		conn, _ = grpc.Dial(r.Scheme()+":///round_robin", grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
+		conn, _ = grpc.Dial(r.Scheme()+":///round_robin", grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name),
+			grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(grpc_retry.WithMax(o.MaxRetry))))
 	} else {
 		// It does not return error if the collector is not running
-		conn, _ = grpc.Dial(o.CollectorHostPort[0], grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
+		conn, _ = grpc.Dial(o.CollectorHostPort[0], grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name),
+			grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(grpc_retry.WithMax(o.MaxRetry))))
```
Comment (yurishkuro) [inline cmd/agent/app/reporter/grpc/collector_proxy.go]: Can we DRY this so that only the actual differences are inside the if branches?

Comment (yurishkuro) [reply cmd/agent/app/reporter/grpc/collector_proxy.go]: Also, let's format it one option per line.

File: cmd/agent/app/reporter/grpc/reporter.go [hunk 2]
```diff
@@ -29,6 +29,10 @@ import (
 	"github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
 )
 
+const (
+	defaultMaxRetry = 3
+)
```
Comment (yurishkuro) [inline cmd/agent/app/reporter/grpc/reporter.go]: Can we move this to flags.go where it is used?

File: cmd/agent/app/reporter/grpc/collector_proxy.go [hunk 3]
```diff
@@ -41,19 +42,23 @@ func NewCollectorProxy(o *Options, mFactory metrics.Factory, logger *zap.Logger)
...
 	} else {
-		// It does not return error if the collector is not running
-		conn, _ = grpc.Dial(o.CollectorHostPort[0], grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
+		target = o.CollectorHostPort[0]
 	}
+	// It does not return error if the collector is not running
+	conn, _ := grpc.Dial(target,
+		grpc.WithInsecure(),
+		grpc.WithBalancerName(roundrobin.Name),
+		grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(grpc_retry.WithMax(o.MaxRetry))))
```
Comment (vprithvi) [inline cmd/agent/app/reporter/grpc/collector_proxy.go]: How does this retry work? Is it possible to configure it with exponential backoff, or some backoff for e.g.?

Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/collector_proxy.go]: By default it is using `BackoffLinearWithJitter`. Which can be changed.

https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/retry/options.go#L26

**Comment by pavolloffay** [comment]
Travis passed, I will try to restart it and hope :)

**Comment by pavolloffay** [comment]
I have restarted the build 4 times, it passed. I will merge this to see how it behaves in other repositories

**Pull request merged**